### DISCUSSION TITLE
pipeline fails with latest_diagonal

### DISCUSSION THREAD

**Discussion by goduckie** [casact/chainladder-python#228] [Q&A]
As per example, a triangle consisting of a latest diagonal only cannot be projected using the pipeline (test 4 below)

```
import chainladder as cl

#A full triangle and a latest_diagonal triangle
raa = cl.load_sample('raa')
raa_latest = raa.latest_diagonal.val_to_dev()

#fit based on full triangle / some benchmark etc
raa_dev_fit = cl.Development().fit(raa)

#test: project triangle to ultimate & dividing by latest diagonal, expect results == cdf's

#test 1 - full: OK 
cl.Chainladder().fit(raa_dev_fit.transform(raa)).ultimate_ / raa_latest.latest_diagonal

#test 2 - latest: OK  
cl.Chainladder().fit(raa_dev_fit.transform(raa_latest)).ultimate_ / raa_latest.latest_diagonal


# test using pipleine
steps = [('sample', 'passthrough'),
         ('dev', raa_dev_fit),
         ('tail', 'passthrough'),
         ('model', cl.Chainladder())]
pipeline = cl.Pipeline(steps)

#test 3 - full: OK
pipeline.fit(raa).named_steps.model.ultimate_ / raa.latest_diagonal

#test 4 - latest: ERROR.  Pipeline fails to project
pipeline.fit(raa_latest).named_steps.model.ultimate_ / raa.latest_diagonal

```

**Comment by goduckie** [comment]
test 4 - can be made to work using:
`pipeline.named_steps.model.fit(pipeline.named_steps.dev.transform(raa_latest)).ultimate_ / raa.latest_diagonal`

is there a way to add the transform step into the pipeline itself?

**Comment by jbogaardt** [comment]
`Pipeline.fit` chains the `fit` and `transform` together for a set of estimators on a **single** triangle, `X`.   In your test 4, `raa_dev_fit` is being refit to just the latest diagonal which doesn't have enough info.  If you want to use the `Pipeline` on a new dataset, you don't need to refit it on the new dataset. The proper usage would be to call `predict` from an existing Pipeline.

```python
#test 4 - Proper usage of Pipeline
pipeline.predict(raa_latest).ultimate_ / raa.latest_diagonal
```

**Comment by goduckie** [comment]
Thank you for the explanation @jbogaardt 

That makes sense.

Will re-run the test and hopefully close out the issue.

**Comment by goduckie** [comment]
To clarify, before one can run the `pipeline.predict` method, one needs to have cl.Chainladder().fit(raa_latest) in the pipeline, Else predict fails.

```

steps = [('sample', 'passthrough'),
         ('dev', raa_dev_fit),
         ('tail', 'passthrough'),
         ('model', cl.Chainladder().fit(raa_latest))]

cl.Pipeline(steps).predict(raa_latest).ultimate_  / raa.latest_diagonal
```

**Reply by jbogaardt** [reply]
This setup seems confusing to me.  None of the steps of the `Pipeline` requires separate fitting either before or inside the Pipeline declaration.  A streamlined examples would look like this:

```python
import chainladder as cl

raa = cl.load_sample('raa')
raa_latest = raa.latest_diagonal

pipe = cl.Pipeline(steps = [
    ('sample', 'passthrough'),
    ('dev', cl.Development()),
    ('tail', 'passthrough'),
    ('model', cl.Chainladder())]
)

pipe.fit(raa)
pipe.predict(raa_latest).ultimate_ / raa.latest_diagonal
```

**Reply by goduckie** [reply]
Agree that is more streamlined. In the workflow I'm considering cl.Development().fit() is created & used earlier in the process, hence it is being inserted directly in the pipeline rather than being refitted, but perhaps that needs a rethink.
 
Thanks for your feedback
